### ISSUE TITLE
mssql+pyodbc error with database using _SC collation

### ISSUE THREAD

**Issue by gordthompson** [sqlalchemy/sqlalchemy#5651]
Follow-up to #5649 and https://github.com/mkleehammer/pyodbc/issues/835

Cannot insert strings longer than 2000 characters into an nvarchar(max) column when the database uses an "_SC" ([Unicode] Supplementary Characters) collation.

Repro code, tested against 87c24c49

``` python
import urllib

import pyodbc
import sqlalchemy as sa


def get_connection_string(db_name):
    return (
        r"DRIVER=ODBC Driver 17 for SQL Server;"
        r"SERVER=(local)\SQLEXPRESS;"
        f"DATABASE={db_name};"
        r"Trusted_Connection=Yes;"
        r"UseFMTONLY=Yes;"
    )


db_name = "gh_sqla_5651"
table_name = "table1"

cnxn = pyodbc.connect(get_connection_string("master"), autocommit=True)

cnxn.execute(f"DROP DATABASE IF EXISTS {db_name}")
cnxn.execute(f"CREATE DATABASE {db_name} COLLATE Latin1_General_100_CI_AI_SC")
cnxn.execute(f"USE {db_name}")
cnxn.execute(
    f"CREATE TABLE {table_name} (id int PRIMARY KEY, txt nvarchar(max))"
)
cnxn.close()

sqlalchemy_uri = "mssql+pyodbc:///?odbc_connect=" + urllib.parse.quote_plus(
    get_connection_string(db_name)
)
engine = sa.create_engine(sqlalchemy_uri)
table1 = sa.Table(table_name, sa.MetaData(), autoload_with=engine)
with engine.begin() as conn:
    conn.execute(table1.insert(), {"id": 1, "txt": "x" * 2001})
"""console output:
sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42000', 
"[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]
Cannot convert to text/ntext or collate to 'Latin1_General_100_CI_AI_SC' 
because these legacy LOB types do not support the Unicode supplementary 
characters whose codepoints are U+10000 or greater. Use types varchar(max), 
nvarchar(max), or a collation which does not have the _SC flag. (4189) 
(SQLParamData)")
"""
```

**Comment by zzzeek** [comment]
I have driver version 13, if i run the test like below I dont get an error.    still using CREATE DATABASE though?

```
import urllib

import pyodbc
import sqlalchemy as sa


db_name = "gh_sqla_5651"
table_name = "table1"


engine = sa.create_engine("mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/test?driver=ODBC+Driver+13+for+SQL+Server", echo=True)

cnxn = engine.raw_connection()
cnxn.rollback()
cnxn.connection.autocommit = True
cnxn.execute(f"DROP DATABASE IF EXISTS {db_name}")
cnxn.execute(f"CREATE DATABASE {db_name} COLLATE Latin1_General_100_CI_AI_SC")
cnxn.execute(f"USE {db_name}")
cnxn.execute(
    f"CREATE TABLE {table_name} (id int PRIMARY KEY, txt nvarchar(max))"
)
cnxn.commit()
cnxn.invalidate()

table1 = sa.Table(table_name, sa.MetaData(), autoload_with=engine, schema="%s.dbo" % db_name)
with engine.begin() as conn:
    conn.execute(table1.insert(), {"id": 1, "txt": "x" * 2001})

```

**Comment by zzzeek** [comment]
OH WEIRD, I get the error if I connect to that DB directly!!! what's the diifference!!!!   

error:

```
engine = sa.create_engine("mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/%s?driver=ODBC+Driver+13+for+SQL+Server" % db_name, echo=True)

table1 = sa.Table(table_name, sa.MetaData(), autoload_with=engine)

```

no error:

```
engine = sa.create_engine("mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/test?driver=ODBC+Driver+13+for+SQL+Server", echo=True)

table1 = sa.Table(table_name, sa.MetaData(), autoload_with=engine, schema="%s.dbo" % db_name)

```

same table!  how is that working?

**Comment by zzzeek** [comment]
OK can confirm the approach in #5649 works, so we need to document the recipe for this situation:

```
@sa.event.listens_for(engine, "do_setinputsizes")
def do_setinputsizes(
    inputsizes, cursor, statement, parameters, context
):
    for key in inputsizes:
        if isinstance(key.type, sa.NVARCHAR):
            inputsizes[key] = (pyodbc.SQL_WVARCHAR, 2005, None)

```

the length needs to be present and needs to accommodate for the size of text being sent.   this text is present in "parameters" if people want to look for it.

this is *absolutely* something pyodbc, or some new extension to pyodbc that simply handles all the known SQL Server typing issues, should be doing.    pyodbc is written in C and they can do things like this very efficiently.  for us it means we'd have to iterate through all the parameters, which could be tens of thousands of lists, and do len() on every value to find the length.   not happening

**Comment by zzzeek** [comment]
also, *very* suspicious I can INSERT into this table from a different database w/ different collation.  there has to be an easier way to do this.

**Comment by gordthompson** [comment]
@zzzeek - My first thought was that this was just an unfortunate legacy behaviour of the msodbcsql17 driver but [now I'm not so sure](https://github.com/mkleehammer/pyodbc/issues/835#issuecomment-711068011).

**Comment by gordthompson** [comment]
Waiting for a response from pyodbc

https://github.com/mkleehammer/pyodbc/issues/835

**Comment by gordthompson** [comment]
[Feedback from pyodbc](https://github.com/mkleehammer/pyodbc/issues/835#issuecomment-712928009) suggests that it is triggering the error because it uses a mechanism called DAE (data-at-execution) when binding the parameter values while the other environments listed above apparently do not. Furthermore, the concern is that

> while msodbcsql doesn't require the use of DAE for the long types, other drivers may.

[I suggested](https://github.com/mkleehammer/pyodbc/issues/835#issuecomment-712969159) that pyodbc add a kwarg to the `pyodbc.connect()` function, e.g.,

``` python
cnxn = pyodbc.connect(connection_string, avoid_dae_on_execute=True)
```

so we could integrate that into the `mssql+pyodbc` dialect (perhaps optionally) when establishing the DBAPI connection.

**Comment by zzzeek** [comment]
afaict this is a dupe of #8661 and was fixed in f4214975a7deb5e13f8b6cf21e39697821396 .   the code works as of that commit.
